### PR TITLE
WIP: Add bitwise operations (XOR, OR, AND)

### PR DESCRIPTION
**Pull request by ralf1307**
I would like to implement a check in CodeGenerator.ts to check for those new operators and if they are used inside a condition, as students could try to use '&' instead of '&&' or '|' instead of '||' in a condition.

### CODE REVIEW

File: src/client/compiler/parser/Parser.ts [hunk 1]
```diff
@@ -11,12 +11,14 @@ type ASTNodes = ASTNode[];
...
     static operatorPrecedence: TokenType[][] = [Parser.assignmentOperators,
     [TokenType.ternaryOperator], [TokenType.colon],
-    [TokenType.or], [TokenType.and],
-    [TokenType.keywordInstanceof, TokenType.lower, TokenType.lowerOrEqual, TokenType.greater, TokenType.greaterOrEqual, TokenType.equal, TokenType.notEqual],
+    [TokenType.or], [TokenType.and], [TokenType.OR], [TokenType.XOR], [TokenType.AND],
+    [TokenType.equal, TokenType.notEqual],
+    [TokenType.keywordInstanceof, TokenType.lower, TokenType.lowerOrEqual, TokenType.greater, TokenType.greaterOrEqual],
+    [TokenType.shiftLeft, TokenType.shiftRight, TokenType.shiftRightUnsigned],
     [TokenType.plus, TokenType.minus], [TokenType.multiplication, TokenType.division, TokenType.modulo]
     ];
```
Comment (ralf1307) [inline src/client/compiler/parser/Parser.ts:23]: I changed the order a bit to more closely relate to [java's operator precedence](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/operators.html)

File: src/client/compiler/parser/Parser.ts [hunk 2]
```diff
@@ -1037,6 +1039,7 @@ export class Parser {
     }
 
     // -, not, this, super, a.b.c[][].d, a.b(), b() (== this.b()), super.b(), super()
+    // FIXME: @RALF add tilde support
```
Comment (ralf1307) [inline src/client/compiler/parser/Parser.ts:1042]: That's one of the reasons this MR is still WIP.

**Review by ralf1307 — COMMENTED** [review]
Should I seperate this commit into multiple ones (Lexer, Parser, PrimitiveTypes)?
Or should I keep it squashed?

File: src/client/compiler/parser/Parser.ts [hunk 3]
```diff
@@ -2084,12 +2087,12 @@ export class Parser {
...
             sextends = this.parseType();
         }
-
+        /*
         while (this.comesToken(TokenType.ampersand)) {
             this.nextToken(); // Skip ampersand
             simplements.push(this.parseType());
         }
-
+        */
```
[the remark this answers is not in the document]
Comment (martin-pabst) [reply src/client/compiler/parser/Parser.ts:2095]: Hello Ralph,
this code fragment makes bounded type parameters inside generic types possible, see [hier, section "multiple bounds"](https://docs.oracle.com/javase/tutorial/java/generics/bounded.html). We need to keep it.

File: src/client/compiler/parser/Parser.ts (hunk 2 above)
Comment (martin-pabst) [reply src/client/compiler/parser/Parser.ts:1042]: See Parser.parseUnary()

**Comment by martin-pabst** [comment]
Hello Ralph,
thanks for this excellent pull request! I merged it by hand and added support or tilde-operator. 
I haven't done any testing yet. -> try to finalize it this weekend.

**Pull request closed without merging**